File: src/main/java/org/janelia/saalfeldlab/paintera/config/CrosshairConfig.java
package org.janelia.saalfeldlab.paintera.config;

import java.util.Arrays;
import java.util.Collection;

import javafx.beans.property.BooleanProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ObservableBooleanValue;
import javafx.scene.paint.Color;
import org.janelia.saalfeldlab.paintera.ui.Crosshair;
import org.janelia.saalfeldlab.util.Colors;

public class CrosshairConfig
{

	public static final Color DEFAULT_ON_FOCUS_COLOR = Colors.CREMI;

	public static final Color DEFAULT_OUT_OF_FOCUS_COLOR = Color.WHITE.deriveColor(0, 1, 1, 0.5);

	private final ObjectProperty<Color> onFocusColor = new SimpleObjectProperty<>(DEFAULT_ON_FOCUS_COLOR);

	private final ObjectProperty<Color> outOfFocusColor = new SimpleObjectProperty<>(DEFAULT_OUT_OF_FOCUS_COLOR);

	private final BooleanProperty showCrosshairs = new SimpleBooleanProperty(true);

	private final BooleanProperty wasChanged = new SimpleBooleanProperty(false);

	{
		onFocusColor.addListener((obs, oldv, newv) -> wasChanged.set(true));
		outOfFocusColor.addListener((obs, oldv, newv) -> wasChanged.set(true));
		showCrosshairs.addListener((obs, oldv, newv) -> wasChanged.set(true));
		wasChanged.addListener((obs, oldv, newv) -> wasChanged.set(false));
	}

	public Color getOnFocusColor()
	{
		return onFocusColor.get();
	}

	public void setOnFocusColor(final Color color)
	{
		this.onFocusColor.set(color);
	}

	public ObjectProperty<Color> onFocusColorProperty()
	{
		return this.onFocusColor;
	}

	public Color getOutOfFocusColor()
	{
		return outOfFocusColor.get();
	}

	public void setOutOfFocusColor(final Color color)
	{
		this.outOfFocusColor.set(color);
	}

	public ObjectProperty<Color> outOfFocusColorProperty()
	{
		return this.outOfFocusColor;
	}

	public boolean getShowCrosshairs()
	{
		return showCrosshairs.get();
	}

	public void setShowCrosshairs(final boolean show)
	{
		this.showCrosshairs.set(show);
	}

	public BooleanProperty showCrosshairsProperty()
	{
		return this.showCrosshairs;
	}

	public ObservableBooleanValue wasChanged()
	{
		return this.wasChanged();
	}

	public void bindCrosshairsToConfig(final Collection<Crosshair> crosshairs)
	{
		crosshairs.stream().forEach(this::bindCrosshairToConfig);
	}

	public void bindCrosshairsToConfig(final Crosshair... crosshairs)
	{
		this.bindCrosshairsToConfig(Arrays.asList(crosshairs));
	}

	public void bindCrosshairToConfig(final Crosshair crosshair)
	{
		crosshair.highlightColorProperty().bind(this.onFocusColor);
		crosshair.regularColorProperty().bind(this.outOfFocusColor);
		crosshair.isVisibleProperty().bind(this.showCrosshairs);
	}
}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/config/CrosshairConfig.kt
package org.janelia.saalfeldlab.paintera.config

import javafx.beans.property.BooleanProperty
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ObservableBooleanValue
import javafx.scene.paint.Color
import org.janelia.saalfeldlab.paintera.ui.Crosshair
import org.janelia.saalfeldlab.util.Colors

class CrosshairConfig {

    private val onFocusColor = SimpleObjectProperty(DEFAULT_ON_FOCUS_COLOR)

    private val outOfFocusColor = SimpleObjectProperty(DEFAULT_OUT_OF_FOCUS_COLOR)

    private val showCrosshairs = SimpleBooleanProperty(true)

    private val wasChanged = SimpleBooleanProperty(false)

    init {
        onFocusColor.addListener { _, _, _ -> wasChanged.set(true) }
        outOfFocusColor.addListener { _, _, _ -> wasChanged.set(true) }
        showCrosshairs.addListener { _, _, _ -> wasChanged.set(true) }
        wasChanged.addListener { _, _, _ -> wasChanged.set(false) }
    }

    fun getOnFocusColor(): Color {
        return onFocusColor.get()
    }

    fun setOnFocusColor(color: Color) {
        this.onFocusColor.set(color)
    }

    fun onFocusColorProperty(): ObjectProperty<Color> {
        return this.onFocusColor
    }

    fun getOutOfFocusColor(): Color {
        return outOfFocusColor.get()
    }

    fun setOutOfFocusColor(color: Color) {
        this.outOfFocusColor.set(color)
    }

    fun outOfFocusColorProperty(): ObjectProperty<Color> {
        return this.outOfFocusColor
    }

    fun getShowCrosshairs(): Boolean {
        return showCrosshairs.get()
    }

    fun setShowCrosshairs(show: Boolean) {
        this.showCrosshairs.set(show)
    }

    fun showCrosshairsProperty(): BooleanProperty {
        return this.showCrosshairs
    }

    fun wasChanged(): ObservableBooleanValue {
        return this.wasChanged
    }

    fun bindCrosshairsToConfig(crosshairs: Collection<Crosshair>) {
        crosshairs.stream().forEach { this.bindCrosshairToConfig(it) }
	}

    fun bindCrosshairsToConfig(vararg crosshairs: Crosshair) {
        this.bindCrosshairsToConfig(listOf(*crosshairs))
    }

    fun bindCrosshairToConfig(crosshair: Crosshair) {
        crosshair.highlightColorProperty().bind(this.onFocusColor)
        crosshair.regularColorProperty().bind(this.outOfFocusColor)
        crosshair.isVisibleProperty.bind(this.showCrosshairs)
    }

    companion object {

        val DEFAULT_ON_FOCUS_COLOR = Colors.CREMI

        val DEFAULT_OUT_OF_FOCUS_COLOR = Color.WHITE.deriveColor(0.0, 1.0, 1.0, 0.5)
    }
}
